core/src/main/java/org/apache/spark/shuffle/sort/BypassMergeSortShuffleWriter.java

```diff
@@ -120,6 +124,13 @@
     this.writeMetrics = writeMetrics;
     this.serializer = dep.serializer();
     this.shuffleExecutorComponents = shuffleExecutorComponents;
+    if ((boolean) conf.get(package$.MODULE$.SHUFFLE_CHECKSUM())) {
+      this.checksumEnabled = true;
+      this.partitionChecksums = new Adler32[numPartitions];
+      for (int i = 0; i < numPartitions; i ++) {
+        this.partitionChecksums[i] = new Adler32();
```

Comment:
> Also, this will need to be conveyed to ESS (filename extension ?) and allow to be configurable in spark.

hmm..I'm not sure if get this concern. Could you elaborate more?

Reply:
The checksum algorithm being used is specified at ESS and spark executor as `Adler32` in code currently - from an evolution point of view, if we have to change/support other checksum algo's in future, this will become an incompatible change : which will require ESS and client to be upgraded together to support it.

Thoughts ?

Reply:
yeah I agree it would be nice to be configurable and recorded.  perhaps either extension on checksum file or metadata in the checksum file.   I would expect ESS to indicate if its supported - ie perhaps error is unknown or unsupported checksum type when trying to diagnose.   I think the other part might be in the push based shuffle. I assume  if its merging files it may have to recalculate these so needs to know the algorithm to use?

Reply:
+CC @otterc for @tgravescs's on push based shuffle.

Reply:
Yes, for push-based shuffle the server would need to know the which algorithm to use. Just thinking out loud, the server can use this checksum to verify that a block which was pushed to it was corrupt or not by calculating and comparing the checksum. 
- This verification may be expensive though while the blocks are getting merged. So, maybe we can think of doing it asynchronously.
- Also may require adding checksum to block push message which would be backward incompatible so maybe we create a new message for it.

Anyways, so if we do want the algorithm to be configurable, can we leverage the `RegisterExecutor` message for it?
We can sent the additional information about what algorithm is being used and if the shuffle server doesn't support it, checksum is not calculated. This would be similar to how we were sending the attemptId information to the shuffle server for push-based shuffle.

Just a thought.

Reply:
> Anyways, so if we do want the algorithm to be configurable, can we leverage the RegisterExecutor message for it?

The reason I was initially proposing adding checksum algo to the file name itself as a suffix is to minimize the state required to reason about which algorithm is being used. We wont need to pass it from container to ESS, or persist it across ESS restarts, etc.

Tom's additional suggestion of including it in checksum file itself as metadata also works - given the current index'ing into the file for a given partition (8 * partition_id), metadata at end of file might be more convenient place to record this.

Reply:
The reason I suggested to leverage the `RegistorExecutor` message to communicate the checksum algo to ESS is because for push-based shuffle, the method of communicating the checksum algorithm to the ESS via checksum file, which is generated when the shuffle data is generated, will not work. This is if we want to validate that a pushed block is corrupt or not at the remote ESS in future. 

However, leveraging this for push-based shuffle is not part of this change but since it was mentioned I thought how push-based shuffle can utilize it.

Reply:
Excellent point - this would handle the case of merger not having any local shuffle files, but only data pushed to it.